cpp/src/arrow/builder.cc

```diff
@@ -873,6 +900,16 @@ Status DictionaryBuilder<T>::Finish(std::shared_ptr<Array>* out) {
   return Status::OK();
 }
 
+Status DictionaryBuilder<NullType>::Finish(std::shared_ptr<Array>* out) {
+  std::shared_ptr<Array> dictionary = std::make_shared<NullArray>(0);
```

Comment:
So one nuance with categorical data that we will need to address at some point (not necessarily right now) -- null is a valid category, so in some applications the result of a dictionary builder for null type should have a dictionary with a single value. We will need to address this more broadly across the dictionary builder implementations, so perhaps let's open a JIRA and look into it later.